client/mm/mm.go

```diff
@@ -73,6 +201,513 @@ func duplicateBotConfig(cfgs []*BotConfig) error {
 	return nil
 }
 
+func priceOracleFromConfigs(ctx context.Context, cfgs []*BotConfig, log dex.Logger) (*priceOracle, error) {
+	var oracle *priceOracle
+	var err error
+	marketsRequiringOracle := marketsRequiringPriceOracle(cfgs)
+	if len(marketsRequiringOracle) > 0 {
+		oracle, err = newPriceOracle(ctx, marketsRequiringOracle, log)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create PriceOracle: %v", err)
+		}
+	}
+
+	return oracle, nil
+}
+
+func (m *MarketMaker) loginAndUnlockWallets(pw []byte, cfgs []*BotConfig) error {
+	err := m.core.Login(pw)
+	if err != nil {
+		return fmt.Errorf("failed to login: %w", err)
+	}
+	unlocked := make(map[uint32]interface{})
+	for _, cfg := range cfgs {
+		if _, done := unlocked[cfg.BaseAsset]; !done {
+			err := m.core.OpenWallet(cfg.BaseAsset, pw)
+			if err != nil {
+				return fmt.Errorf("failed to unlock wallet for asset %d: %w", cfg.BaseAsset, err)
+			}
+			unlocked[cfg.BaseAsset] = true
+		}
+
+		if _, done := unlocked[cfg.QuoteAsset]; !done {
+			err := m.core.OpenWallet(cfg.QuoteAsset, pw)
+			if err != nil {
+				return fmt.Errorf("failed to unlock wallet for asset %d: %w", cfg.QuoteAsset, err)
+			}
+			unlocked[cfg.QuoteAsset] = true
+		}
+	}
+
+	return nil
+}
+
+func validateAndFilterEnabledConfigs(cfgs []*BotConfig) ([]*BotConfig, error) {
+	enabledCfgs := make([]*BotConfig, 0, len(cfgs))
+	for _, cfg := range cfgs {
+		if !cfg.Disabled {
+			enabledCfgs = append(enabledCfgs, cfg)
+		}
+	}
+	if len(enabledCfgs) == 0 {
+		return nil, errors.New("no enabled bots")
+	}
+	if err := duplicateBotConfig(enabledCfgs); err != nil {
+		return nil, err
+	}
+	return enabledCfgs, nil
+}
+
+func (m *MarketMaker) setupBalances(cfgs []*BotConfig) error {
+	m.botBalances = make(map[string]*botBalances, len(cfgs))
+
+	type trackedBalance struct {
+		balanceAvailable uint64
+		balanceReserved  uint64
+	}
+
+	balanceTracker := make(map[uint32]*trackedBalance)
+
+	for _, cfg := range cfgs {
+		if _, found := balanceTracker[cfg.BaseAsset]; !found {
+			bal, err := m.core.AssetBalance(cfg.BaseAsset)
+			if err != nil {
+				return fmt.Errorf("failed to get balance for asset %d: %v", cfg.BaseAsset, err)
+			}
+
+			balanceTracker[cfg.BaseAsset] = &trackedBalance{
+				balanceAvailable: bal.Available,
+			}
+		}
+
+		if _, found := balanceTracker[cfg.QuoteAsset]; !found {
+			bal, err := m.core.AssetBalance(cfg.QuoteAsset)
+			if err != nil {
+				return fmt.Errorf("failed to get balance for asset %d: %v", cfg.QuoteAsset, err)
+			}
+			balanceTracker[cfg.QuoteAsset] = &trackedBalance{
+				balanceAvailable: bal.Available,
+			}
+		}
+
+		baseBalance := balanceTracker[cfg.BaseAsset]
+		quoteBalance := balanceTracker[cfg.QuoteAsset]
+
+		var baseRequired, quoteRequired uint64
+		if cfg.BaseBalanceType == Percentage {
+			baseRequired = baseBalance.balanceAvailable * cfg.BaseBalance / 100
+		} else {
+			baseRequired = cfg.BaseBalance
+		}
```

Comment:
Do we need to validate the balance types across `[]*BotConfig` somehow? What if the user is trading on a couple markets with the same base asset, and they give `Percentage`  values that sum to > 100? What happens if they provide mixed `Percentage` and `Amount` balance types for the same asset?

Reply:
This is being validated in `setupBalances`.

Reply:
Ah. I guess you mean because it would trigger an error below at 
```Go
if baseRequired > baseBalance.balanceAvailable-baseBalance.balanceReserved { ...
```
I still wonder if it wouldn't be easy to screw up. As an example, say the balance of xyz is 2. We run one bot that reserves 1 xyz, and one bot that reserves 50% of xyz. If the balance of xyz changes to 1.5 for whatever reason, then those allocations cannot both be satisfied. 

OK for now, but just wanted to point it out.